Implement post and pre hooks for optimizer

anupambhatnagar opened the pull request:
Fixes #88446


cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @VitalyFedyunin @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen

anupambhatnagar committed:
first commit

anupambhatnagar committed:
fix flake8 issues

robieta commented:
An easy way to test the hooks is to pass a callback that records some state. So something like:
```
optimizer = model.parameters()

logged = []

def callback(*args, **kwargs):
    logged.append(None)
    
optimizer.step()
self.assertEqual(len(logged), 0)

register_hook(callback)
optimizer.step()
self.assertEqual(len(logged), 1)
```

etc.

anupambhatnagar committed:
call pre and post hooks

albanD commented on torch/optim/optimizer.py:
```diff
@@ -42,9 +65,14 @@ class Optimizer(object):
     def __init__(self, params, defaults):
         torch._C._log_api_usage_once("python.optimizer")
         self.defaults = defaults
+        super().__setattr__('_optimizer_step_pre_hooks', OrderedDict())
```
nit: no need to use super() here. You can just do `self._optimizer_step_pre_hooks = OrderedDict()`.

albanD commented on torch/optim/optimizer.py:
```diff
@@ -137,8 +175,15 @@ def wrapper(*args, **kwargs):
                 with torch.autograd.profiler.record_function(profile_name):
+                    # call optimizer step pre hooks
+                    for hook in self._optimizer_step_pre_hooks.values():
+                        pre_hook_result = self._pre_hook(hook)
```
```python
maybe_new_closure = hook(self, closure)
if maybe_new_closure is not None:
    closure = maybe_new_closure
```

albanD commented:
I think for the function names I would either:
- Rename the monkey patch function to reflect that it is not profiler-specific anymore.
- Use a second independent monkey patching function (similar to the profiler one) and then we will remove the profiler one in the future.

albanD commented:
/easycla

anupambhatnagar committed:
adding test

anupambhatnagar committed:
clean the test code

anupambhatnagar committed:
updating tests

anupambhatnagar committed:
clean tests

anupambhatnagar committed:
uncomment assert statement

anupambhatnagar committed:
fix lint issues

anupambhatnagar committed:
adding tests to test_optim.py

anupambhatnagar committed:
fix tests

anupambhatnagar committed:
Merge branch 'master' into optimizer-hooks

anupambhatnagar committed:
resolve conflicts

anupambhatnagar committed:
fix github action errors

anupambhatnagar committed:
fix typo in post hook test

anupambhatnagar committed:
refactor tests into different files

anupambhatnagar committed:
cleanup

anupambhatnagar committed:
more cleanup

anupambhatnagar committed:
Merge branch 'master' into optimizer-hooks

anupambhatnagar committed:
update comments and public API function names

albanD commented on torch/optim/optimizer.py:
```diff
@@ -27,6 +34,19 @@ def _use_grad(self, *args, **kwargs):
     return _use_grad
 
 
+def register_optimizer_step_pre_hook(hook: Callable[..., None]) -> RemovableHandle:
```
We should remove these if they just return an error and do nothing.

anupambhatnagar replied:
will do.

anupambhatnagar commented on test/test_optimizer_hooks.py:
```diff
@@ -0,0 +1,92 @@
+    @classmethod
+    def setUpClass(cls):
+        params = [torch.Tensor([1, 1])]
+        cls.optimizer = SGD(params, lr=0.001)
```
we need a unique SGD instance per class. If it that is not the case, then the first test passes and remaining test fails.

anupambhatnagar commented on test/test_optimizer_hooks.py:
```diff
@@ -0,0 +1,92 @@
+            data.append(1)
+
+        data = []
+        pre_handles, post_handles = [], []
```
In this test both pre hook and post hook are called. different lists are being used to store pre/post hook handles.
I have `data` as a list to ensure that pre/post hook work on something other than integers as in the first two tests.

anupambhatnagar commented on test/test_optimizer_hooks.py:
```diff
@@ -0,0 +1,92 @@
+        self.optimizer.step()
+        self.optimizer.step()
+        # check if pre hooks were registered
+        assert data == 6
```
will do.

anupambhatnagar commented on torch/optim/optimizer.py:
```diff
@@ -127,8 +156,7 @@ def _optimizer_step_code(self):
 
-    def _hook_for_profile(self):
-        self._zero_grad_profile_name = "Optimizer.zero_grad#{}.zero_grad".format(self.__class__.__name__)
+    def _step_with_hook(self):
```
sure

anupambhatnagar commented on torch/optim/optimizer.py:
```diff
@@ -137,17 +165,75 @@ def wrapper(*args, **kwargs):
 
             return wrapper
 
+        self._zero_grad_profile_name = "Optimizer.zero_grad#{}.zero_grad".format(self.__class__.__name__)
```
to improve readability - all code of the parent function now is positioned together rather than spread out before and after the `profile_hook_step` function. I can revert it if you think the original presentation is easier to read.

anupambhatnagar commented on test/test_optimizer_hooks.py:
```diff
@@ -0,0 +1,92 @@
+from unittest import TestCase
+
+class TestOptimizerHook(TestCase):
+    optimizer: SGD
```
probably not needed. will remove it.

anupambhatnagar committed:
address PR comments

anupambhatnagar committed:
add global optimizer hooks and fix tests

anupambhatnagar commented:
@albanD I have fixed the issue with the tests. The issue was that `obj` in `profile_hook_step` was being misused for `self`. I have refactored the code `profile_hook_step` based on @robieta suggestions and now the `profile_hook_step` has been converted into a static function of the Optimizer class instead of being a closure. I'm placing back the test cases in `test_optim.py`. Additionally I have added global optimizer hook implementation and now the test checking both pre/post hooks simultaneously contains two optimizers to check that global optimizer hooks are being called properly.

Big thanks to @robieta for helping me root cause the issue with the test cases.

anupambhatnagar commented:
@albanD I ssh'd into the server and ran the same test that is failing in CI. Upon manually running `pytest test/test_optim.py` all tests passed. Here's a screenshot.
![Screen Shot 2022-11-30 at 9 22 57 PM](https://user-images.githubusercontent.com/91992329/204972361-2cbee176-0aba-4ebf-8490-0fe9d8f534b5.png)

albanD commented on test/test_optim.py:
```diff
@@ -1558,6 +1560,101 @@ def test_empty_grad(self):
                 self.assertEqual(original_param, param)
 
 
+    def test_post_hook(self):
```
The test seems to fail only on the dynamo shard. You can add @skipIfTorchDynamo on these tests and open an issue with module: dynamo for the right people to look into this.

albanD reviewed (commented):
SGTM
Just skip the dynamo test to make CI green and then that should be ready to land!

anupambhatnagar committed:
adding skipIfTorchDynamo on tests

albanD reviewed (approved):
Sounds good! Thanks for the update!

albanD commented:
@pytorchbot merge -g
